### PR TITLE
refactor metrics

### PR DESCRIPTION
**Pull request by fcakyon**
changes:
- simplify code by directly parsing validation/logging frequency from pytorch lightning trainer
- metrics tests are more explicit
- 94% coverage on metrics tests
- reformat with `python -m tests.run_code_style format`

usage:

```python
from torchrush.loggers import TensorboardLogger, NeptuneLogger
from torchrush.metrics import MetricCallback

metric_callback = MetricCallback(metrics=['accuracy', 'f1', 'precision', 'recall'])

trainer = pl.Trainer(
    max_epochs=10,
    check_val_every_n_epoch=1,
    val_check_interval=1.0,  # set as 0.33 if you want to calculate 3 validation per epoch
    logger=[TensorboardLogger(), NeptuneLogger()],
    callbacks=[metric_callback]
)
```

### CODE REVIEW

**Comment by fcakyon** [comment]
@devrimcavusoglu reformat yapinca base'de 1 satirlik degisiklik yapti, sorun olmaz ins 😅

File: torchrush/metrics.py [hunk 1]
```diff
@@ -77,7 +72,7 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
 
         pl_module.log_hyperparams()
 
-    def _accumulate_metrics(self, outputs, mode: str = "train"):
+    def _add_batch(self, outputs, mode: str = "train"):
```
Comment (devrimcavusoglu) [inline torchrush/metrics.py]: Type hint ekleyebiliriz `outputs: Dict[str, Any]`.

File: torchrush/metrics.py [hunk 2]
```diff
@@ -121,12 +116,11 @@ def on_train_batch_end(
...
         batch_idx: int,
     ) -> None:
-        # accumulate metrics
-        self._accumulate_metrics(outputs, mode="train")
+        self._add_batch(outputs, mode="train")
 
         # log metrics
         pl_module.log_any({"train/loss": outputs["loss"].item()}, step=batch_idx)
-        if batch_idx % self.eval_freq == 0 and self.log_on == "batch_end":
+        if not trainer.sanity_checking and trainer.fit_loop.epoch_loop._should_check_val_fx():
```
Comment (devrimcavusoglu) [inline torchrush/metrics.py]: Bu statement'ı bir (private) methoda dönüştürebiliriz, `_should_log()` veya `_should_log_metrics()` gibi (daha iyi isimler olabilir benim aklım şu an gelenler bunlar 😄), bi çok methodda kullanmışız.

File: torchrush/metrics.py [hunk 3]
```diff
@@ -155,22 +148,9 @@ def on_test_batch_end(
         batch_idx: int,
         dataloader_idx: int,
     ) -> None:
-        # accumulate metrics
-        self._accumulate_metrics(outputs, mode="test")
+        self._add_batch(outputs, mode="test")
```
Comment (devrimcavusoglu) [inline torchrush/metrics.py:154]: Burada benim şöyle bi çekincem vardı, kendim deneme yaparken, `add_batch`, `compute` call olduktan sonra sıfırlıyor sanırım accumulate ettiği pred/refleri, bunla alakalı bi sıkıntı yoksa ignore edebilirsin bu commenti.

File: tests/torchrush/test_metrics.py [hunk 4]
```diff
@@ -86,6 +86,184 @@ def batch_step_outputs():
...
+    pl_trainer = pl.Trainer(
+        enable_checkpointing=False,
+        max_steps=100,
+        val_check_interval=val_check_interval,
+        logger=[mock_logger],
+        callbacks=[metric_callback],
+    )
+
+    # attach dataloaders and model to trainer
+    train_loader, val_loader = data_loaders
```
Comment (devrimcavusoglu) [inline tests/torchrush/test_metrics.py]: 139-151 (inclusive) aralığını bir fonksiyona dönüştürüp bu checki orada yapabiliriz, `logger_sanity_check` gibi bişey olabilir 3 fixture'da da aynı şekilde kullanmışız (152'yi ayrı tutuyorum), code duplicationdan kaçınalım olabildiğince.

File: torchrush/metrics.py (hunk 3 above)
Comment (fcakyon) [reply torchrush/metrics.py:154]: evet her compute cagirisimizda sifirliyor diye biliyorum. bu bir sorun mu bizim durumumuzda? 

ben epochta 1 cagirmayi dusundugum icin her epoch sonunda sifirlamasini istiyorum zaten.

gercek bir deneyde test edip istedigimizi karsiliyor mu diye bakabilirsin.

File: torchrush/metrics.py [hunk 5]
```diff
@@ -72,7 +72,7 @@ def on_train_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule")
 
         pl_module.log_hyperparams()
 
-    def _add_batch(self, outputs, mode: str = "train"):
+    def _add_batch(self, outputs: Dict[str, torch.Tensor], mode: str = "train"):
```
Comment (devrimcavusoglu) [inline torchrush/metrics.py]: `Tensor` olma zorunluluğu olmamalı bence, pratik bir örnek NLP tarafı için olabilir mesela, iki List of strings karşılaştırıyoruz. Yani NLP case'i için logitleri hesaplayıp daha sonra decode edebiliriz. [Implementasyon](https://github.com/devrimcavusoglu/torchrush-dev/blob/6c381b74ffefb15a6d239c72f8caec3b61a65b8f/torchrush/metrics.py#L87). Bu sebeple `Dict[str, Any]` olmalı bence bu kullanıma yönelik. Eğer `Any` çok broad dersen, belki `Union[torch.Tensor, np.ndarray, List]` gibi bişey de olabilir value kısmı.

**Review by devrimcavusoglu — APPROVED** [review]

File: torchrush/metrics.py (hunk 5 above)
Comment (devrimcavusoglu) [reply torchrush/metrics.py]: Bu arda `torch.Tensor` string itemlara izin vermiyor diye biliyorum, veriyosa bile çok iyi bi pratik değil. Buna en yakın `np.array` içinde string tutmak olabilir ama normaldeki common practice `list` olarak kullanım.

Comment (fcakyon) [reply torchrush/metrics.py]: @devrimcavusoglu hocam sen girdi olarak ne verebilelim istiyorsun? tensor + list of text desteği eklesek yeterli olur mu?

Comment (devrimcavusoglu) [reply torchrush/metrics.py]: biz destek olarak bişey eklemeyelim, kullanıcı biz outputs'u dict olarak verilmesini istiyoruz, bu dict içinde `"loss"` dışında istediğini verebilir, yani

```python
outputs = {
  "loss": torch.Tensor(0.3523)
  "predictions": np.array([[0,1,2,0], [0,0,1,1]])
}
```
gibi bişey de verebilir, biz `Dict[str, torch.Tensor]` yaparak sadece torch.Tensor verebilirsin bu dict'te demeyelim, o yüzden de önerim sadece type hinti değiştirmek, `Dict[str, Any]` şeklinde.

Comment (fcakyon) [reply torchrush/metrics.py]: @devrimcavusoglu lossun tensor oldugunu varsayarak .item() ile python degiskenine ceviriyoruz mesela: https://github.com/devrimcavusoglu/torchrush-dev/blob/d08a6211f2f82296d753e58b8f609f37886f3f6f/torchrush/metrics.py#L141

Eger loss degerinin List[str, int, float] olarak da girilmesini destekleyeceksek bir check eklememiz gerekmez mi?

```python

if type(loss).__module__ in ["torch", "numpy"]:
    pl_module.log_any(outputs["loss"].item())
else:
    pl_module.log_any(outputs["loss"])

```

Gerci `log_any` metodu tensor/float/string hepsini handle edebiliyordu degil mi? o zaman sorun olmaz 🤔

**Comment by fcakyon** [comment]
type hintleri guncelledim 👍

File: torchrush/metrics.py (hunk 5 above)
Comment (devrimcavusoglu) [reply torchrush/metrics.py]: hmm, evet concern'u şimdi anladım, aslında dediğin gibi loss için type'lar belli ve tensor olmalı dediğin gibi, ama bu sadece `loss` için, mesela `"predictions"` gibi bir key için tensor olmasını zorlamamalıyız, dediğinde haklısın ama bence bu şekilde daha uygun, "loss"un valuesu için tensor/float olması biraz implicit ama orayı pytorch lightning handle etmeli artık, biz belki docstringe ekleyebiliriz ileride.

**Review by devrimcavusoglu — APPROVED** [review]

**Comment by devrimcavusoglu** [comment]
LGTM 👍

File: torchrush/metrics.py (hunk 5 above)
Comment (fcakyon) [reply torchrush/metrics.py]: bu haliyle yanlis anlamadiysam loss'un tum turlerde olmasini destekliyoruz, torch/numpy/float
